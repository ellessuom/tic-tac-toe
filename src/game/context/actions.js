export const PLAYERS_ID = {
  PLAYER_ONE: 'p1',
  PLAYER_TWO: 'p2',
};

export const INIT = {
  canPlay: true, // allows to add new items to the matrix
  isTie: false, // no winners - all tiles filled
  currentPlayer: PLAYERS_ID.PLAYER_ONE, // user id, either 'p1' or 'p2'
  usedTiles: []
};

export default (state, dispatch) => {
  return ({
    startGame: () => {
      dispatch({
        ...INIT,
        [PLAYERS_ID.PLAYER_ONE]: {
          ...state[PLAYERS_ID.PLAYER_ONE],
          selectedTiles: []
        },
        [PLAYERS_ID.PLAYER_TWO]: {
          ...state[PLAYERS_ID.PLAYER_TWO],
          selectedTiles: []
        }
      });
    },
    play: (tileIdx) => {
      dispatch({
        usedTiles: [...state.usedTiles, tileIdx].sort(),
        [state.currentPlayer]: {
          ...state[state.currentPlayer],
          selectedTiles: [...state[state.currentPlayer].selectedTiles, tileIdx].sort(),
        },
        currentPlayer: state.currentPlayer === PLAYERS_ID.PLAYER_TWO? PLAYERS_ID.PLAYER_ONE : PLAYERS_ID.PLAYER_TWO
      });
    },
  });
};
