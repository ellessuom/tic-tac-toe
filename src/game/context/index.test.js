import React from 'react';
import { shallow } from 'enzyme';
import * as DataContext from './index';
import { INIT } from "./actions";

const TILE_IDX = 3;
describe('<BasePlayer />', () => {
  let wrapper;
  let actions;
  beforeAll(() => {
    wrapper = shallow(
      <DataContext.Provider>
        <div id="child-test"></div>
      </DataContext.Provider>
    );
    actions = wrapper.props().value[1];
  });

  it('should contain test child', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it ('should render with correct INIT data', () => {
    expect(wrapper.props().value[0]).toMatchObject(INIT);
  });

  describe('Simulate play', () => {
    beforeAll(() => {
      actions.play(TILE_IDX);
    });

    it('should have update data with correct next player', () => {
      const data = wrapper.props().value[0];
      expect(data.currentPlayer).toBe('p2');
    });

    it('should have update data with correct tile index', () => {
      const data = wrapper.props().value[0];
      expect(data.usedTiles).toEqual([TILE_IDX]);
    });

    it('should have updated player used tiles with correct tile index', () => {
      const data = wrapper.props().value[0];
      expect(data.p1.selectedTiles).toEqual([TILE_IDX]);
    });

    it('should have not updated the other player with the selected tile', () => {
      const data = wrapper.props().value[0];
      expect(data.p2.selectedTiles).toEqual([]);
    });
  });

  describe('Simulate restart', () => {
    beforeAll(() => {
      actions.play(TILE_IDX);
      actions.startGame();
    });

    it('expect to have reset the game', () => {
      expect(wrapper.props().value[0]).toMatchObject(INIT);
    });
  });
});
