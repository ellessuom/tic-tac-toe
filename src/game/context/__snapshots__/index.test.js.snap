// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasePlayer /> should contain test child 1`] = `
<ContextProvider
  value={
    Array [
      Object {
        "canPlay": true,
        "currentPlayer": "p1",
        "isTie": false,
        "p1": Object {
          "selectedTiles": Array [],
          "wins": 0,
        },
        "p2": Object {
          "selectedTiles": Array [],
          "wins": 0,
        },
        "usedTiles": Array [],
        "winner": "",
        "winnerPattern": "",
      },
      Object {
        "play": [Function],
        "startGame": [Function],
        "verify": [Function],
      },
    ]
  }
>
  <div
    id="child-test"
  />
</ContextProvider>
`;
